File: src/poller.rs
use anyhow::Result;
use reqwest::Client;
use std::time::Duration;
use tokio::time;

use crate::db::Db;
use crate::fetcher::fetch_feed;

pub async fn run_once(db: Db, client: Client) -> Result<()> {
    // 全フィードを取得
    let feeds = {
        let db2 = db.clone();
        tokio::task::spawn_blocking(move || db2.get_feed_targets()).await??
    };

    for target in feeds {
        let db_clone = db.clone();
        let client_clone = client.clone();
        if let Err(e) = fetch_feed(
            db_clone,
            client_clone,
            target.id,
            target.url.clone(),
            target.etag,
            target.last_modified,
        )
        .await
        {
            tracing::warn!("Failed to fetch feed {}: {}", target.url, e);
        }
    }

    // FTSインデックス再構築
    {
        let db2 = db.clone();
        tokio::task::spawn_blocking(move || {
            if let Err(e) = db2.rebuild_search_index() {
                tracing::warn!("FTS rebuild failed: {}", e);
            }
        })
        .await?;
    }

    Ok(())
}

pub async fn start_poller(db: Db, client: Client, interval_minutes: u64) {
    // 起動直後に1回実行
    let db2 = db.clone();
    let client2 = client.clone();
    tokio::spawn(async move {
        tracing::info!("Initial poll starting...");
        if let Err(e) = run_once(db2, client2).await {
            tracing::warn!("Initial poll error: {}", e);
        }
        tracing::info!("Initial poll done");
    });

    // 定期ポーリング
    tokio::spawn(async move {
        let mut interval = time::interval(Duration::from_secs(interval_minutes * 60));
        interval.tick().await; // 最初のティックは即時なのでスキップ
        loop {
            interval.tick().await;
            tracing::info!("Scheduled poll starting...");
            if let Err(e) = run_once(db.clone(), client.clone()).await {
                tracing::warn!("Scheduled poll error: {}", e);
            }
        }
    });
}

File: internal/poller/poller.go
package poller

import (
	"log/slog"
	"math"
	"net/http"
	"time"

	"github.com/cross-ts/rss-reader/internal/db"
	"github.com/cross-ts/rss-reader/internal/fetcher"
	"github.com/cross-ts/rss-reader/internal/opmlsync"
)

// maxIntervalMinutes is the largest poll interval (in minutes) that can be
// safely converted to a time.Duration without overflow.
const maxIntervalMinutes = uint64(math.MaxInt64 / int64(time.Minute))

// defaultIntervalMinutes is the fallback interval used when the caller
// provides a value that would overflow the time.Duration conversion.
const defaultIntervalMinutes = 15

// RunOnce fetches all feed targets and applies the results.
func RunOnce(database *db.DB, client *http.Client) error {
	targets, err := database.GetFeedTargets()
	if err != nil {
		return err
	}

	slog.Info("poller: starting run", "feed_count", len(targets))

	for _, t := range targets {
		result, err := fetcher.FetchFeedData(client, &t)
		if err != nil {
			slog.Warn("poller: fetch feed failed", "url", t.URL, "error", err)
			continue
		}
		if result == nil {
			// Not modified (304).
			continue
		}
		inserted, err := database.ApplyFetchResult(t.ID, result.Articles, result.Meta)
		if err != nil {
			slog.Warn("poller: apply fetch result failed", "url", t.URL, "error", err)
			continue
		}
		if inserted > 0 {
			slog.Info("poller: new articles", "url", t.URL, "count", inserted)
		}
	}

	if err := database.RebuildSearchIndex(); err != nil {
		return err
	}

	slog.Info("poller: run complete")
	return nil
}

// Start launches the background poller that runs immediately and then on each tick.
func Start(database *db.DB, client *http.Client, intervalMinutes uint64, syncer *opmlsync.Syncer) {
	if intervalMinutes > maxIntervalMinutes {
		slog.Warn("poller: interval minutes out of range, falling back to default", "interval_minutes", intervalMinutes, "default", defaultIntervalMinutes)
		intervalMinutes = defaultIntervalMinutes
	}
	interval := time.Duration(intervalMinutes) * time.Minute

	// Run immediately in a goroutine.
	go func() {
		if syncer != nil {
			if err := syncer.SyncIfChanged(); err != nil {
				slog.Warn("poller: opml sync failed", "error", err)
			}
		}
		if err := RunOnce(database, client); err != nil {
			slog.Error("poller: initial run failed", "error", err)
		}
	}()

	// Start ticker goroutine.
	go func() {
		ticker := time.NewTicker(interval)
		defer ticker.Stop()

		for range ticker.C {
			if syncer != nil {
				if err := syncer.SyncIfChanged(); err != nil {
					slog.Warn("poller: opml sync failed", "error", err)
				}
			}
			if err := RunOnce(database, client); err != nil {
				slog.Error("poller: scheduled run failed", "error", err)
			}
		}
	}()
}
